### PR TITLE
[search] Separated out ranker.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

File: search/geocoder.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #include "search/features_layer_matcher.hpp"
 #include "search/house_numbers_matcher.hpp"
 #include "search/locality_scorer.hpp"
+#include "search/pre_ranker.hpp"
```
Comment (ygorshenin) [inline search/geocoder.cpp]: Why do we need pre_ranker here?

File: search/processor.hpp [hunk 2]
```diff
@@ -53,15 +56,14 @@ struct QueryParams;
 class ReverseGeocoder;
 
 class Geocoder;
+class Ranker;
+// todo(@m) Merge with Ranker.
```
Comment (ygorshenin) [inline search/processor.hpp]: s/todo/TODO/g

File: search/processor.hpp [hunk 3]
```diff
@@ -128,17 +130,22 @@ class Processor : public my::Cancellable
 
   friend string DebugPrint(ViewportID viewportId);
 
-  friend class impl::FeatureLoader;
-  friend class impl::BestNameFinder;
-  friend class impl::PreResult2Maker;
-  friend class impl::DoFindLocality;
-  friend class impl::HouseCompFactory;
+  friend class FeatureLoader;
```
Comment (ygorshenin) [inline search/processor.hpp]: Sort, please.

File: search/processor.hpp [hunk 4]
```diff
@@ -208,9 +208,9 @@ class Processor : public my::Cancellable
 
 protected:
   bool m_viewportSearch;
-  bool m_keepHouseNumberInQuery;
 
   PreRanker m_preRanker;
+  unique_ptr<Ranker> m_ranker;
```
Comment (ygorshenin) [inline search/processor.hpp]: Why do we need unique_ptr here?

File: search/processor.hpp [hunk 5]
```diff
@@ -128,17 +130,22 @@ class Processor : public my::Cancellable
...
-  void ForEachCategoryTypes(StringSliceBase const & slice, ToDo toDo) const;
-  template <class ToDo>
-  void ProcessEmojiIfNeeded(strings::UniString const & token, size_t ind, ToDo & toDo) const;
+
+  void ForEachCategoryType(
+      StringSliceBase const & slice,
+      function<void(size_t /* tokenId */, uint32_t /* typeId */)> const & fn) const;
+
+  void ProcessEmojiIfNeeded(
+      strings::UniString const & token, size_t ind,
```
Comment (ygorshenin) [inline search/processor.hpp]: s/ind/index

or

s/ind/ix

File: search/ranker.cpp [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+template <typename TSlice>
+void UpdateNameScore(vector<strings::UniString> const & tokens, TSlice const & slice,
+                     NameScore & bestScore)
+{
+  auto const score = GetNameScore(tokens, slice);
+  if (score > bestScore)
+    bestScore = score;
+}
+
+void RemoveDuplicatingLinear(vector<IndexedValue> & indV)
```
Comment (ygorshenin) [inline search/ranker.cpp]: s/indV/values/g

File: search/ranker.cpp [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    search::RankingInfo info;
+    InitRankingInfo(ft, center, res1, info);
+    info.m_rank = NormalizeRank(info.m_rank, info.m_searchType, center, country);
+    res2->SetRankingInfo(move(info));
+
+    return res2;
+  }
+};
+
+bool Ranker::IsResultExists(PreResult2 const & p, vector<IndexedValue> const & indV)
```
Comment (ygorshenin) [inline search/ranker.cpp]: s/indV/values/g

File: search/ranker.hpp [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+class Ranker
+{
+public:
+  Ranker(PreRanker & preRanker, Processor & processor)
+    : m_viewportSearch(false), m_preRanker(preRanker), m_processor(processor)
+  {
+  }
+
+  void Init(bool viewportSearch) { m_viewportSearch = viewportSearch; }
```
Comment (ygorshenin) [inline search/ranker.hpp]: inline?

File: search/ranker.hpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  void FlushViewportResults(Geocoder::Params const & params, Results & res);
+
+private:
+  bool m_viewportSearch;
+  PreRanker & m_preRanker;
+
+  // todo(@m) Remove.
+  Processor & m_processor;
+};
+
```
Comment (ygorshenin) [inline search/ranker.hpp]: Remove this blank line, please.

File: search/ranker.hpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  Result MakeResult(PreResult2 const & r) const;
+
+  void FlushResults(Geocoder::Params const & params, Results & res, size_t resCount);
+  void FlushViewportResults(Geocoder::Params const & params, Results & res);
+
+private:
+  bool m_viewportSearch;
+  PreRanker & m_preRanker;
+
+  // todo(@m) Remove.
```
Comment (ygorshenin) [inline search/ranker.hpp]: s/todo/TODO/g

File: search/geocoder.cpp (hunk 1 above)
Comment (mpimenov) [inline search/geocoder.cpp]: Well we should include it somewhere, shouldn't we? Note that it is not included in geocoder.hpp

File: search/processor.hpp (hunk 2 above)
Comment (mpimenov) [inline search/processor.hpp]: Sorry, but not in this CL.

File: search/processor.hpp (hunk 3 above)
Comment (mpimenov) [inline search/processor.hpp]: done

File: search/processor.hpp (hunk 4 above)
Comment (mpimenov) [inline search/processor.hpp]: What to you propose?

File: search/geocoder.cpp (hunk 1 above)
Comment (ygorshenin) [inline search/geocoder.cpp]: Is it used in geocoder?

File: search/processor.hpp (hunk 5 above)
Comment (mpimenov) [inline search/processor.hpp]: done

File: search/geocoder.cpp (hunk 1 above)
Comment (ygorshenin) [inline search/geocoder.cpp]: Ah, OK, I see.

File: search/ranker.cpp (hunk 6 above)
Comment (mpimenov) [inline search/ranker.cpp]: done

File: search/ranker.cpp (hunk 7 above)
Comment (mpimenov) [inline search/ranker.cpp]: done

File: search/ranker.hpp (hunk 8 above)
Comment (mpimenov) [inline search/ranker.hpp]: done

File: search/ranker.hpp (hunk 9 above)
Comment (mpimenov) [inline search/ranker.hpp]: done

**Comment by mpimenov** [comment]
PTAL

File: search/processor.hpp (hunk 4 above)
Comment (ygorshenin) [inline search/processor.hpp]: Just write
```
Ranker m_ranker;
```

File: search/ranker.hpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "search/geocoder.hpp"
+#include "search/intermediate_result.hpp"
+#include "search/mode.hpp"
+#include "search/processor.hpp"
```
Comment (ygorshenin) [inline search/ranker.hpp]: You don't need this inclusion as you've forward-declared processor.

File: search/intermediate_result.hpp [hunk 12]
```diff
@@ -153,7 +151,45 @@ inline string DebugPrint(PreResult2 const & t)
...
+    m_distanceToPivot = info.m_distanceToPivot;
+  }
+
+  PreResult2 const & operator*() const { return *m_value; }
+
+  inline double GetRank() const { return m_rank; }
+
+  inline double GetDistanceToPivot() const { return m_distanceToPivot; }
+};
+
```
Comment (ygorshenin) [inline search/intermediate_result.hpp]: Please, remove this blank line.

Comment (mpimenov) [inline search/intermediate_result.hpp]: done

File: search/ranker.hpp (hunk 11 above)
Comment (mpimenov) [inline search/ranker.hpp]: done

File: search/processor.hpp (hunk 4 above)
Comment (mpimenov) [inline search/processor.hpp]: done
Thanks

**Comment by mpimenov** [comment]
PTAL

**Comment by ygorshenin** [comment]
LGTM

**Pull request merged**